plugins/ipam/dhcp/daemon: Remove Unix socket on exit

fnordahl opened the pull request:
Get rid of the need of removing it manually

steveeJ commented on plugins/ipam/dhcp/daemon.go:
```diff
@@ -126,6 +128,16 @@ func getListener() (net.Listener, error) {
+		signal.Notify(signalChan, os.Interrupt)
+		signal.Notify(signalChan, syscall.SIGTERM)
+		go func() {
+			select {
```
If you only have one case there is no need for select

fnordahl commented:
Updated version:
- [x] Removed superflous select {}
- [x] Added functional tests

fnordahl commented:
The coverage test still fails although I have added tests for the code I am adding. Am I being punished from already non-existing tests here? :-) ( @steveeJ )

fnordahl commented:
I guess coveralls does not detect that the code is tested since there are no relevant function calls in the test. I will update with code that simulates the process execution with fork + function call + sleep and then interrupt to see if that makes coveralls happy.

fnordahl commented:
Updated version:
- [x] Use net.Listener.Close() to remove the socket instead of os.Remove()
- [x] Exercise the getListener() function in a test prior to forking of a seperate process for testing the signal processing to make coveralls happy with the code addittion / test coverage ratio.

With this I believe I have addressed the issues that needed rework @steveeJ

jonboulle commented on plugins/ipam/dhcp/daemon.go:
```diff
@@ -126,7 +128,18 @@ func getListener() (net.Listener, error) {
+			go func() {
+				<-signalChan
+				l.Close()
+				os.Exit(1)
```
better to do 128 + signal:
http://tldp.org/LDP/abs/html/exitcodes.html

jonboulle commented on plugins/ipam/dhcp/daemon_test.go:
```diff
@@ -0,0 +1,80 @@
+
+	// verify existence of dhcp socket
+	_, err = os.Stat(socketPath)
+	if err != nil {
```
Better to be a little more discerning when checking stat errors, e.g. for this specific error message the condition should really be `os.IsNotExist(err)` - otherwise, for `err != nil`, a more general message would be appropriate ("error checking socket existence:" or so)

jonboulle commented on plugins/ipam/dhcp/daemon_test.go:
```diff
@@ -0,0 +1,80 @@
+	}
+
+	// wait for process initialization to finish
+	time.Sleep(1 * time.Second)
```
Sleeps are really bad in tests. Is there another way you can wait for the process to initialise? Something you could either block or poll on?

fnordahl commented:
Thanks @jonboulle 

Updated version:
- [x] use 128 + signal as exit code and verify in tests
- [x] make error messages more generic for check on err from os.Stat()
- [x] Don't sleep in tests! Continuously poll for existence of pid file (max 1 second) on start. Use proc.Wait() on quit.

steveeJ commented on plugins/ipam/dhcp/daemon_test.go:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	// verify existence of dhcp socket, keep retrying until 1 second has passed
+	for startTime := time.Now(); err != nil || time.Now().Before(startTime.Add(1*time.Second)); _, err = os.Stat(socketPath) {
```
This loop header looks very confusing. We don't require most compact version of the code, but prefer readability. Please rewrite this accordingly.

steveeJ commented on plugins/ipam/dhcp/daemon_test.go:
```diff
@@ -0,0 +1,95 @@
+	for startTime := time.Now(); err != nil || time.Now().Before(startTime.Add(1*time.Second)); _, err = os.Stat(socketPath) {
+		// NOOP
+	}
+	if err != nil {
```
it should be possible to not have to check `err` again after the loop

steveeJ commented on plugins/ipam/dhcp/daemon_test.go:
```diff
@@ -0,0 +1,95 @@
+		case syscall.Signal:
+			signum = int(signal)
+		default:
+			signum = 0
```
use `signum := 0` instead of the default case?

fnordahl commented:
Thanks @steveeJ 

Updated version:
- [x] Initialize signum before select making the default case redundant
- [x] Re-write retryloop to be more readable, and utilize Go idioms

squeed commented:
The CNI repository has split; if you'd like to re-open this PR against the plugins repository, please do so.
